src/PluginClientBuilder.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[][] List of plugins ordered by priority [priority => Plugin[]]). */
+    private $plugins = [];
+
+    /** @var array Array of options to give to the plugin client */
+    private $options = [];
+
+    /**
+     * @param int $priority Priority of the plugin. The higher comes first.
+     */
+    public function addPlugin(Plugin $plugin, int $priority = 0): self
+    {
+        $this->plugins[$priority][] = $plugin;
+
+        return $this;
+    }
+
+    public function setOption(string $name, $value): self
+    {
+        $this->options[$name] = $value;
+
+        return $this;
+    }
+
+    public function removeOption(string $name): self
+    {
+        unset($this->options[$name]);
+
+        return $this;
+    }
+
+    /**
+     * @param ClientInterface | HttpAsyncClient $client
+     */
+    public function createClient($client): PluginClient
+    {
+        if (!$client instanceof ClientInterface && !$client instanceof HttpAsyncClient) {
+            throw new \RuntimeException('You must provide a valid http client');
+        }
+
+        $plugins = $this->plugins;
+
+        if (0 === count($plugins)) {
+            $plugins[] = [];
+        }
```

Comment:
note, this is to avoid `array_merge(...[])` which provokes an error. Should I remove the `if` and always add an empty array on the next key when creating the plugin client ? Or should I put it in the properties list `[0 => []]` by default ?)

Reply:
(PHPStan is not happy if I do the inverse condition (`if (count($plugins) > 0) ....`), reporting a possible type violation in the plugin cient's constructor.

Reply:
i wonder if we should throw an exception when no plugins are set? if not, i would do `return new PluginClient($client, [], $this->options);` here.

Reply:
If no exceptions are thrown, I don't think it's really worth it to handle this case here if no plugins are given. the `krsort`,`array_merge` and `array_values` on empty array wouldn't really have any impact on perfs, don't you agree ?

Reply:
as we need a sepcial case anyways, i think it would be better to be explicit for better readability. 

oh, and thinking about it: throwing an exception would be weird. if we want that, it should be the PluginClient itself, and i guess we do not want it as the builder might sometimes not register any plugins, if we are in some sort of dynamic setup.